### DISCUSSION TITLE
How can I modify Frame Data directly?

### DISCUSSION THREAD

**Discussion by kurochan** [asticode/go-astiav#79] [Q&A] [answered]
I want to manipulate frame images directly from go.
For example, to overlay other images, add effects, insert generated frames etc....

I know ffmpeg has a filter feature, but this time I want to manipulate the frame directly.

How can I do this?

It seems to me that a function like `Frame.SetData([]byte)` would be a good idea, but if it doesn't exist, is there another way?

**Comment by asticode** [comment]
For now you can "export" the frame's data to a Go image and manipulate it, but you can't "save" your manipulations to a `Frame`. I remember trying to implement exactly that, one or two years ago but couldn't find a clean way while ensuring proper memory management and planes processing. I'll think about it and see if I come up with a new idea 👍

**Comment by asticode** [comment, marked as answer]
@kurochan I've created the branch [frame-data-write](https://github.com/asticode/go-astiav/tree/frame-data-write) with the following:

- `FrameData`.`SetBytes(b []byte, align int) error` which allows you to copy a buffer inside a frame's data
- `FrameData`.`FromImage(src image.Image) error` which allows you to copy a Go `image.Image` inside a frame's data

Bare in mind that the developer is responsible for handling the writability of the frame. For that, I've added

- `Frame`.`MakeWritable()`
- `Frame`.`IsWritable()`

I'm not sure I've implemented the most efficient way to do this but I couldn't think of a better way. Also, memory management shouldn't be a problem, but for those features only time can tell 😬 

Could you try this new branch and let me know whether this work as expected? If so, I'll merge it in `master` 👍

**Reply by kurochan** [reply]
Thank you for the quick implementation. I will try it soon🚀

**Reply by kurochan** [reply]
~Failed to build 🤔~
~> filter_link.go:32:30: l.c.frame_rate undefined (type *_Ctype_AVFilterLink has no field or method frame_rate)~

correction: I had installed ffmpeg n7.1 and modified it to use n7.0, then compilation succeeded.
It appears that some breaking changes were included between n7.0 and n7.1.

**Reply by asticode** [reply]
Are you sure you're using FFMpeg `n7.0`?

**Reply by kurochan** [reply]
Sorry, I was using n7.1 by mistake.

**Reply by asticode** [reply]
No worries, do my changes work properly?

**Reply by kurochan** [reply]
@asticode The compilation succeeds and works, but the picture appears to be partially corrupted.
I wrote code to draw a blue rectangle based on the sample and `go run main.go -i input.ts -o out.ts` .

Could you review that my implementation is correct?
https://gist.github.com/kurochan/083b1cb4fb1439e90a633afa6a7bee5f
Here are the main changes: https://gist.github.com/kurochan/083b1cb4fb1439e90a633afa6a7bee5f#file-draw_avframe-go-L522-L539

![capture](https://github.com/user-attachments/assets/f52e916d-347c-47a3-990b-8b9ff28deda0)

**Reply by asticode** [reply]
As I've stated in my previous comment:

> Bare in mind that the developer is responsible for handling the writability of the frame

FYI I've also added this fact to the documentation in the code 👍 

Your frame's buffer is referenced in multiple places therefore you can't write data manually into it without causing the problems you're seeing. What you need to do is call `f.MakeWritable() error` [here](https://gist.github.com/kurochan/083b1cb4fb1439e90a633afa6a7bee5f#file-draw_avframe-go-L526). After doing so, the content I generated using the code you've shared was correct 👍 

Therefore I'm merging my branch into both `dev` and `master`.

Let me know whether you need a tag 👍

**Reply by kurochan** [reply]
@asticode I understood the points you made. And it worked as intended 🎉  
> Let me know whether you need a tag 👍

Yes, I would like you to release a new version 👍 

Thank you for your quick implementation!

**Reply by asticode** [reply]
I've created a `v0.22.0` tag 👍

**Comment by kurochan** [comment]
Implemented at v0.22.0 🎉
